iam_managed_policy improvements

willthames opened the pull request:
##### SUMMARY
Use snake case for module return values
Add exception handling

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
iam_managed_policy

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel dcd1ff2809) last updated 2017/06/29 13:11:02 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION

Commit:
Change iam_managed_policy to return snake case

Import module_utils.ec2 methods explicitly

Commit:
iam_managed_policy: tidy argument spec

Remove unnecessary specifications (`default=None`, `required=False`)
Use `required_if` to check for `policy` when `state` is `present`

willthames commented on lib/ansible/modules/cloud/amazon/iam_managed_policy.py:
```diff
@@ -163,76 +183,99 @@ def get_or_create_policy_version(iam, policy, policy_document):
+        version = iam.create_policy_version(PolicyArn=policy['Arn'], PolicyDocument=policy_document)['PolicyVersion']
+        return version, True
+    except botocore.exceptions.ClientError as e:
+        delete_oldest_non_default_version(module, iam, policy)
```
@dkhenry I've left the logic here pretty much as it was, but I don't really understand how this section of code is designed to work. Is there a limit to the number of versions allowed (is that what the 5 in your comment refers to?)

dkhenry replied:
Yes there is a limit, it defaults to 5 ,but I am not sure its 5 for everyone as many AWS limits can be changed by a customer, so instead I made the logic to just delete one and try again. That way you are not removing more then you need to in the event a user has a higher cap then I did.

willthames replied:
Thanks for the extra info @dkhenry. I might update the comment slightly.

willthames replied:
I've updated the comments and the error handling.

Commit:
Provide exception handling in iam_managed_policy

Pretty much all AWS API calls can go wrong, and we should handle them.

Update line wrapping to improve readability of method calls

Commit:
Improve error handling when policy version limit exceeded

Better document policy version limit exceeded, and check
the error code to see that that's actually what happened
(rather than e.g. no permissions to add a new version)

willthames commented:
Once this is merged, #26223 can be merged too

ryansb commented on lib/ansible/modules/cloud/amazon/iam_managed_policy.py:
```diff
@@ -102,137 +102,187 @@
-    response = iam.list_policies(Scope='Local', **kwargs)
+def get_policy_by_name(module, iam, name, **kwargs):
+    try:
+        response = iam.list_policies(Scope='Local', **kwargs)
```
`marker` must be capitalized here per http://boto3.readthedocs.io/en/latest/reference/services/iam.html#IAM.Client.list_policies, so line 140 has to call `get_policy_by_name(module, iam, name, Marker=response['Marker'])`

willthames replied:
Good point. Easiest to fix using the paginator with build_full_result, along with AWSRetry

ryansb replied:
Ah, well that'd be even better :+1:

Commit:
iam_managed_policy: better handle pagination and retry
